include/RAJA/exec-agency/scan_agency.hxx

```diff
@@ -0,0 +1,203 @@
+//
+// LLNL-CODE-689114
+//
+// All rights reserved.
+//
+// This file is part of RAJA.
+//
+// For additional details, please also read RAJA/LICENSE.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice,
+//   this list of conditions and the disclaimer below.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the disclaimer (as noted below) in the
+//   documentation and/or other materials provided with the distribution.
+//
+// * Neither the name of the LLNS/LLNL nor the names of its contributors may
+//   be used to endorse or promote products derived from this software without
+//   specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL LAWRENCE LIVERMORE NATIONAL SECURITY,
+// LLC, THE U.S. DEPARTMENT OF ENERGY OR CONTRIBUTORS BE LIABLE FOR ANY
+// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+// OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+// IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "RAJA/config.hxx"
+#include "RAJA/exec-sequential/scan_sequential.hxx"
+#include "RAJA/ThreadUtils_CPU.hxx"
+
+#include "agency/agency.hpp"
+#include "agency/experimental.hpp"
+
+#include <algorithm>
+#include <functional>
+#include <iterator>
+#include <type_traits>
+#include <vector>
+
+namespace RAJA
+{
+namespace detail
+{
+namespace scan
+{
+
+/*!
+        \brief explicit inclusive inplace scan given range, function, and
+   initial value
+*/
+template <typename Iter, typename BinFn, typename ValueT, typename Worker>
+void inclusive_inplace(const ::RAJA::agency_base<agency::parallel_agent, Worker>&,
+                       Iter begin,
+                       Iter end,
+                       BinFn f,
+                       ValueT v)
+{
+  using Value = typename ::std::iterator_traits<Iter>::value_type;
+
+  const int n = end - begin;
+  auto tiles = agency::experimental::tile_evenly(
+      agency::experimental::interval(0, n), getMaxReduceThreadsCPU());
+  const int p = tiles.size();
+
+  ::std::vector<Value> sums(p, v);
+  auto sumsP = &sums[0];
+
+  agency::bulk_invoke(agency::con(p),
```

Comment:
We can't get native barrier support without using `agency::con` and `agency::concurrent_agent` so we sneakily ignore the given policy; it will still be threaded, but with concurrency guarantees. We also lose out on OpenMP, AFAIK there isn't an `agency::omp::con` policy.